File: cpp/src/lance/format/page_table.cc
//  Copyright 2022 Lance Authors
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#include "lance/format/page_table.h"

#include <arrow/builder.h>

#include <memory>
#include <vector>

namespace lance::format {

void PageTable::SetPageInfo(int32_t column_id,
                            int32_t batch_id,
                            int64_t position,
                            int64_t length) noexcept {
  page_info_map_[column_id][batch_id] = std::make_tuple(position, length);
}

/// Get PageInfo
std::optional<PageTable::PageInfo> PageTable::GetPageInfo(int32_t column_id,
                                                          int32_t batch_id) const noexcept {
  auto column_it = page_info_map_.find(column_id);
  if (column_it == page_info_map_.end()) {
    return std::nullopt;
  }
  auto page_it = column_it->second.find(batch_id);
  if (page_it == column_it->second.end()) {
    return std::nullopt;
  }
  return page_it->second;
}

::arrow::Result<int64_t> PageTable::Write(const std::shared_ptr<::arrow::io::OutputStream>& out) {
  ::arrow::Int64Builder builder;

  auto num_columns = page_info_map_.rbegin()->first + 1;
  int32_t num_batches = 0;
  for (auto& [k, m] : page_info_map_) {
    num_batches = std::max(num_batches, m.rbegin()->first + 1);
  }

  ARROW_RETURN_NOT_OK(builder.Reserve(num_columns * num_batches * 2));
  for (int32_t column_id = 0; column_id < num_columns; ++column_id) {
    for (int32_t batch_id = 0; batch_id < num_batches; ++batch_id) {
      auto page_info = GetPageInfo(column_id, batch_id);
      auto [position, length] = page_info.value_or(std::make_tuple(-1, -1));
      ARROW_RETURN_NOT_OK(builder.Append(position));
      ARROW_RETURN_NOT_OK(builder.Append(length));
    }
  }
  ARROW_ASSIGN_OR_RAISE(auto page_table, builder.Finish());
  ARROW_ASSIGN_OR_RAISE(auto pos, out->Tell());
  ARROW_RETURN_NOT_OK(
      out->Write(std::static_pointer_cast<::arrow::Int64Array>(page_table)->values()));
  return pos;
}

::arrow::Result<std::shared_ptr<PageTable>> PageTable::Make(
    const std::shared_ptr<::arrow::io::RandomAccessFile>& in,
    int64_t page_table_position,
    int32_t num_columns,
    int32_t num_batches) {
  ARROW_ASSIGN_OR_RAISE(
      auto buf, in->ReadAt(page_table_position, (num_columns * num_batches * 2 * sizeof(int64_t))));

  auto arr = ::arrow::Int64Array(num_columns * num_batches * 2, buf);

  auto lt = std::make_shared<PageTable>();
  for (int32_t col = 0; col < num_columns; col++) {
    for (int32_t batch = 0; batch < num_batches; batch++) {
      auto idx = col * num_batches + batch;
      auto position = arr.Value(idx * 2);
      auto length = arr.Value(idx * 2 + 1);
      lt->SetPageInfo(col, batch, position, length);
    }
  }
  return lt;
}

}  // namespace lance::format

File: rust/src/format/page_table.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use arrow_array::builder::Int64Builder;
use arrow_array::{Array, Int64Array};
use arrow_schema::DataType;
use std::collections::BTreeMap;
use tokio::io::AsyncWriteExt;

use crate::encodings::plain::PlainDecoder;
use crate::encodings::Decoder;
use crate::error::Result;
use crate::io::object_reader::ObjectReader;
use crate::io::object_writer::ObjectWriter;
use crate::Error;

#[derive(Clone, Debug, PartialEq)]
pub struct PageInfo {
    pub position: usize,
    pub length: usize,
}

impl PageInfo {
    pub fn new(position: usize, length: usize) -> Self {
        Self { position, length }
    }
}

/// Page lookup table.
///
#[derive(Debug, Default)]
pub struct PageTable {
    /// map[field-id,  map[batch-id, PageInfo]]
    pages: BTreeMap<i32, BTreeMap<i32, PageInfo>>,
}

impl PageTable {
    /// Load [PageTable] from disk.
    pub async fn load<'a>(
        reader: &dyn ObjectReader,
        position: usize,
        num_columns: i32,
        num_batches: i32,
    ) -> Result<Self> {
        let length = num_columns * num_batches * 2;
        let decoder = PlainDecoder::new(reader, &DataType::Int64, position, length as usize)?;
        let raw_arr = decoder.decode().await?;
        let arr = raw_arr.as_any().downcast_ref::<Int64Array>().unwrap();

        let mut pages = BTreeMap::default();
        for col in 0..num_columns {
            pages.insert(col, BTreeMap::default());
            for batch in 0..num_batches {
                let idx = col * num_batches + batch;
                let batch_position = &arr.value((idx * 2) as usize);
                let batch_length = &arr.value((idx * 2 + 1) as usize);
                pages.get_mut(&col).unwrap().insert(
                    batch,
                    PageInfo {
                        position: *batch_position as usize,
                        length: *batch_length as usize,
                    },
                );
            }
        }

        Ok(Self { pages })
    }

    pub async fn write(&self, writer: &mut ObjectWriter) -> Result<usize> {
        if self.pages.is_empty() {
            return Err(Error::InvalidInput {
                source: "empty page table".into(),
            });
        }

        let pos = writer.tell();
        let num_columns = self.pages.keys().max().unwrap() + 1;
        let num_batches = self
            .pages
            .values()
            .flat_map(|c_map| c_map.keys().max())
            .max()
            .unwrap()
            + 1;

        let mut builder = Int64Builder::with_capacity((num_columns * num_batches) as usize);
        for col in 0..num_columns {
            for batch in 0..num_batches {
                if let Some(page_info) = self.get(col, batch) {
                    builder.append_value(page_info.position as i64);
                    builder.append_value(page_info.length as i64);
                } else {
                    builder.append_slice(&[0, 0]);
                }
            }
        }
        let arr = builder.finish();
        writer
            .write_all(arr.into_data().buffers()[0].as_slice())
            .await?;

        Ok(pos)
    }

    /// Set page lookup info for a page identified by `(column, batch)` pair.
    pub fn set(&mut self, column: i32, batch: i32, page_info: PageInfo) {
        self.pages
            .entry(column)
            .or_insert_with(BTreeMap::default)
            .insert(batch, page_info);
    }

    pub fn get(&self, column: i32, batch: i32) -> Option<&PageInfo> {
        self.pages.get(&column).and_then(|c_map| c_map.get(&batch))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_set_page_info() {
        let mut page_table = PageTable::default();
        let page_info = PageInfo::new(1, 2);
        page_table.set(10, 20, page_info.clone());

        let actual = page_table.get(10, 20).unwrap();
        assert_eq!(actual, &page_info);
    }
}
